src/main/ruby/core/stat.rb

```diff
@@ -57,45 +73,56 @@ class Stat
 
     attr_reader :path
 
-    def initialize(path)
-      path = Rubinius::Type.coerce_to_path(path)
-      result = Truffle.invoke_primitive(:stat_stat, self, path)
-      Errno.handle_jnr path unless result == 0
+    def initialize(path_or_buffer)
+      if path_or_buffer.is_a?(Buffer)
+        @buffer = path_or_buffer.to_h
+      else
+        path = Rubinius::Type.coerce_to_path(path_or_buffer)
+        Buffer.new do |buffer|
+          result = Truffle::POSIX.truffleposix_stat(path, buffer)
+          Errno.handle_nfi path unless result == 0
+          @buffer = buffer.to_h
+        end
+      end
     end
 
     def self.stat(path)
       path = Rubinius::Type.coerce_to_path(path)
-      stat = allocate
-      if Truffle.invoke_primitive(:stat_stat, stat, path) == 0
-        stat
-      else
-        nil
+      Buffer.new do |buffer|
+        result = Truffle::POSIX.truffleposix_stat(path, buffer)
+        if result == 0
+          return Stat.new buffer
+        else
+          return nil
+        end
       end
     end
 
     def self.lstat(path)
       stat = lstat?(path)
-      Errno.handle_jnr path unless stat
+      Errno.handle_nfi path unless stat
       stat
     end
 
     def self.lstat?(path)
       path = Rubinius::Type.coerce_to_path(path)
-      stat = allocate
-      result = Truffle.invoke_primitive(:stat_lstat, stat, path)
-      if result == 0
-        stat
-      else
-        nil
+      Buffer.new do |buffer|
+        result = Truffle::POSIX.truffleposix_lstat(path, buffer)
+        if result == 0
+          return Stat.new buffer
+        else
+          return nil
+        end
       end
     end
 
     def self.fstat(fd)
       fd = Rubinius::Type.coerce_to fd, Integer, :to_int
-      stat = allocate
-      result = Truffle.invoke_primitive(:stat_fstat, stat, fd)
-      Errno.handle_jnr "file descriptor #{descriptor}" unless result == 0
-      stat
+      Buffer.new { |buffer|
+        result = Truffle::POSIX.truffleposix_fstat(fd, buffer)
+        Errno.handle_nfi "file descriptor #{descriptor}" unless result == 0
+        return Stat.new buffer
+      }
```

Comment:
This is another case where we use `do`...`end`. Not a big deal, but something to keep an eye out for.

Reply:
Fixing.